R/rdeephaven/R/client_wrapper.R

```diff
@@ -1,110 +1,249 @@
+      options$add_int_option(new_int_option[1], as.numeric(new_int_option[2]))
     }
-  )
+
+    if (string_option != "") {
+      verify_string("string_option", string_option, TRUE)
+      new_string_option <- strsplit(string_option, ":", fixed = TRUE)
+      options$add_string_option(new_string_option[1], new_string_option[2])
+    }
+
+    if (extra_header != "") {
+      verify_string("extra_header", extra_header, TRUE)
+      new_extra_header <- strsplit(extra_header, ":", fixed = TRUE)
+      options$add_extra_header(new_extra_header[1], new_extra_header[2])
+    }
+
+    internal_client <- new(INTERNAL_Client,
+      target = target,
+      client_options = options
+    )
+    return(new("Client", .internal_rcpp_object = internal_client))
+  }
+)
+
+### HELPER FUNCTIONS ###
+# These functions return RC objects returned by Rcpp without wrapping them in S4
+
+check_for_table <- function(client, name) {
+  return(client@.internal_rcpp_object$check_for_table(name))
+}
+
+rbr_to_dh_table <- function(client, rbr) {
+  ptr <- client@.internal_rcpp_object$new_arrow_array_stream_ptr()
+  rbr$export_to_c(ptr)
+  return(client@.internal_rcpp_object$new_table_from_arrow_array_stream_ptr(ptr))
+}
+
+arrow_to_dh_table <- function(client, arrow_tbl) {
+  rbr <- as_record_batch_reader(arrow_tbl)
+  return(rbr_to_dh_table(client, rbr))
+}
+
+tibble_to_dh_table <- function(client, tibbl) {
+  arrow_tbl <- arrow_table(tibbl)
+  return(arrow_to_dh_table(client, arrow_tbl))
+}
+
+df_to_dh_table <- function(client, data_frame) {
+  arrow_tbl <- arrow_table(data_frame)
+  return(arrow_to_dh_table(client, arrow_tbl))
+}
+
+### USER-FACING METHODS ###
+
+setGeneric(
+  "open_table",
+  function(client_instance, name) {
+    return(standardGeneric("open_table"))
+  },
+  signature = c("client_instance", "name")
+)
+
+#' @export
+setMethod(
+  "open_table",
+  signature = c(client_instance = "Client"),
+  function(client_instance, name) {
+    verify_string("name", name, TRUE)
+    if (!check_for_table(client_instance, name)) {
+      stop(paste0("The table '", name, "' you're trying to pull does not exist on the server."))
+    }
+    return(new("TableHandle", .internal_rcpp_object = client_instance@.internal_rcpp_object$open_table(name)))
+  }
+)
+
+setGeneric(
+  "empty_table",
+  function(client_instance, size) {
+    return(standardGeneric("empty_table"))
+  },
+  signature = c("client_instance", "size")
+)
+
+#' @export
+setMethod(
+  "empty_table",
+  signature = c(client_instance = "Client"),
+  function(client_instance, size) {
+    verify_positive_int("size", size, TRUE)
+    return(new("TableHandle", .internal_rcpp_object = client_instance@.internal_rcpp_object$empty_table(size)))
+  }
+)
+
+setGeneric(
+  "time_table",
+  function(client_instance, ...) {
+    return(standardGeneric("time_table"))
+  },
+  signature = "client_instance"
+)
+
+#' @export
+setMethod(
+  "time_table",
+  signature = c(client_instance = "Client"),
```

Comment:
Not sure what the dispatch signature would be in this optional arg case. 
... how will this evolve as strings are possible.  Another dispatch question.

Reply:
I think this can evolve with more generics to add more types without impacting the API at all.